File: benchmarks/C++_Primitives/Sieve.cpp
#include <iostream>
#include <string>
#include <btypes_primitives/BUtils.hpp>
#include <btypes_primitives/BSet.hpp>
#include <btypes_primitives/BInteger.hpp>
#include <btypes_primitives/BBoolean.hpp>

#ifndef Sieve_H
#define Sieve_H

using namespace std;

class Sieve {

    public:



    private:




        BSet<BInteger > numbers;
        BInteger cur;
        BInteger limit;

    public:

        Sieve() {
            numbers = (BSet<BInteger>::interval((BInteger(2)),(BInteger(1000000))));
            cur = (BInteger(2));
            limit = (BInteger(1000000));
        }

        BInteger ComputeNumberOfPrimes() {
            BInteger res;
            while(((BBoolean(cur.greater((BInteger(1))).booleanValue() && cur.multiply(cur).lessEqual(limit).booleanValue()))).booleanValue()) {
                if((numbers.elementOf(cur)).booleanValue()) {
                    BInteger n;
                    BSet<BInteger > set;
                    n = cur;
                    set = (BSet<BInteger >());
                    while((n.lessEqual(limit.divide(cur))).booleanValue()) {
                        set = set._union((BSet<BInteger >(cur.multiply(n))));
                        n = n.plus((BInteger(1)));
                    }
                    numbers = numbers.difference(set);
                } 
                cur = cur.plus((BInteger(1)));
            }
            res = numbers.card();
            return res;
        }

};

int main() {
    Sieve sieve;
    cout << sieve.ComputeNumberOfPrimes();
    return 0;
}

#endif

File: benchmarks/Java_Clojure_Set_Primitives/Sieve.java
import de.hhu.stups.btypes.BSet;
import de.hhu.stups.btypes.BInteger;
import de.hhu.stups.btypes.BBoolean;
import java.util.Objects;
import de.hhu.stups.btypes.BUtils;

public class Sieve {






    private BSet<BInteger> numbers;
    private BInteger cur;
    private BInteger limit;

    public Sieve() {
        numbers = BSet.interval(new BInteger(2),new BInteger(1000000));
        cur = new BInteger(2);
        limit = new BInteger(1000000);
    }

    public BInteger ComputeNumberOfPrimes() {
        BInteger res = null;
        while((new BBoolean(cur.greater(new BInteger(1)).booleanValue() && cur.multiply(cur).lessEqual(limit).booleanValue())).booleanValue()) {
            if((numbers.elementOf(cur)).booleanValue()) {
                BInteger n = null;
                BSet<BInteger> set = null;
                n = cur;
                set = new BSet<BInteger>();
                while((n.lessEqual(limit.divide(cur))).booleanValue()) {
                    set = set.union(new BSet<BInteger>(cur.multiply(n)));
                    n = n.plus(new BInteger(1));
                }
                numbers = numbers.difference(set);
            } 
            cur = cur.plus(new BInteger(1));
        }
        res = numbers.card();
        return res;
    }

    public static void main(String[] args) {
        Sieve sieve = new Sieve();
        sieve.ComputeNumberOfPrimes();
    }


}
